### PR TITLE
Display code language

### PR DESCRIPTION
**Pull request by Littlegolden**
#3 display the code language on the top right corner of `<pre>`

### CODE REVIEW

File: js/src/forum/helpers/hljsDefaults.js [hunk 1]
```diff
@@ -393,9 +393,43 @@ export function hljsDefaults(theme, baseUrl) {
 
     var x = document.getElementsByTagName('pre');
     for(var i=0; i<x.length; ++i)
-        x[i].style = 'background: ' + background + ';';
+        x[i].style = 'background: ' + background + ';' + 'color: ' + color;
```
Comment (Littlegolden) [inline js/src/forum/helpers/hljsDefaults.js:396]: this gives the color of `pre:after` element

File: js/src/forum/helpers/hljsDefaults.js [hunk 2]
```diff
@@ -393,9 +393,43 @@ export function hljsDefaults(theme, baseUrl) {
...
+            }
+        });
+    });
+
+    elements.forEach(function(element){
+        observer.observe(element, {
+            attributes: true,
+            attributeFilter: ['class']
+        });
+    })
```
Comment (Littlegolden) [inline js/src/forum/helpers/hljsDefaults.js:417]: Asynchronous loading. Add listener to `<pre>` tag, after its class attribute changed,  execute the `show_language` method.

File: js/src/forum/helpers/hljsDefaults.js [hunk 3]
```diff
@@ -393,9 +393,43 @@ export function hljsDefaults(theme, baseUrl) {
...
+
+    var pres = document.getElementsByTagName('pre');    
+    var code_language;
+    function show_language(){
+        for(var i=0; i<pres.length; ++i) {
+            if(pres[i].className==""){
```
Comment (Littlegolden) [inline js/src/forum/helpers/hljsDefaults.js:423]: if haven't give code language to `<pre>`, we give it, else we skip.

File: js/src/forum/helpers/hljsDefaults.js [hunk 4]
```diff
@@ -393,9 +393,43 @@ export function hljsDefaults(theme, baseUrl) {
...
+                try {
+                    code_language = pres[i].getElementsByTagName('code').item(0).result.language;
+                    if(code_language != "properties"){
+                        pres[i].classList.add("language-" + code_language);
+                    }
+                } catch (error) {
+                    continue;
+                }
```
Comment (Littlegolden) [inline js/src/forum/helpers/hljsDefaults.js]: Incase this code could not work if the `language` attribute can not be found,

**Comment by Littlegolden** [comment]
#3

**Comment by therealsujitk** [comment]
Thanks for taking this up! I'm a bit busy right now, but I'll test this locally as soon as I'm free.

**Comment by Littlegolden** [comment]
And I have add the copy code function, it's testing at https://discuss.flarum.org.cn/d/2478/2

**Comment by Littlegolden** [comment]
I separated #2 #3 into the clipboardjs extension, as I found merging them together was not very easy to maintain.

**Pull request closed without merging**